### PR TITLE
Update lint deprecation for tool lints

### LINKED ISSUE #4349: Deprecation of lints does not work correctly with tool lints
Observed this while working on #4348 

1. `util/dev update_lints` incorrectly removes the lint from `src/lintlist/mod.rs`. It should be marked as `deprecated` instead by using the `deprecation` field
2. The lint deprecation does not take the `clippy::` prefix into account

### PR DESCRIPTION
**Pull request by phansch**
changelog: Allow tool lints (`clippy::*`) to be deprecated

Our lint deprecation previously didn't work for tool lints, because
`register_removed` was registering lints to be removed _without_ the
`clippy` prefix.

Fixes #4349

### CODE REVIEW

**Comment by phansch** [comment]
Note that the current change means, that this will stop deprecation warnings when using non-tool lints, like `#[allow(assign_ops)]`. There will just be an `unknown_lint` warning for non-tool lints.

cc @flip1995

**Comment by flip1995** [comment]
> Note that the current change means, that this will stop deprecation warnings when using non-tool lints

When I implemented the tool lints, I was aware of this problem, but decided to leave it untouched until we heave to deprecate another lint. The reasoning behind that was, that people who still use the deprecated lints, probably won't use tool lints. That reasoning still applies, so the best case would be to allow deprecating both tool lints and non-tool lints. But IIRC that wasn't that easy to implement (otherwise I would have done it (probably 😄)).

I think every maintained crate swapped to tool lints by now and we're on the safe side to make this change?

**Comment by phansch** [comment]
> allow deprecating both tool lints and non-tool lints.

I haven't tried it, but wouldn't it be enough to register both the tool version and the non-tool version as removed?

**Comment by flip1995** [comment]
> There will just be an unknown_lint warning for non-tool lints.

This warning is not optimal, but I think as long as there is _some_ warning about deprecated lints, this should be fine.

File: clippy_lints/src/lib.rs [hunk 1]
```diff
@@ -387,47 +387,47 @@ pub fn register_plugins(reg: &mut rustc_plugin::Registry<'_>, conf: &Conf) {
...
     );
     store.register_removed(
-        "unstable_as_mut_slice",
+        "clippy::unstable_as_mut_slice",
         "`Vec::as_mut_slice` has been stabilized in 1.7",
     );
     store.register_removed(
-        "str_to_string",
+        "clippy::str_to_string",
         "using `str::to_string` is common even today and specialization will likely happen soon",
```
Comment (flip1995) [inline clippy_lints/src/lib.rs]: "specialization will likely happen soon" 15e55f5d lol 😄

File: clippy_lints/src/lib.rs [hunk 2]
```diff
@@ -387,47 +387,47 @@ pub fn register_plugins(reg: &mut rustc_plugin::Registry<'_>, conf: &Conf) {
...
     );
     store.register_removed(
-        "assign_ops",
+        "clippy::assign_ops",
         "using compound assignment operators (e.g., `+=`) is harmless",
     );
     store.register_removed(
-        "if_let_redundant_pattern_matching",
+        "clippy::if_let_redundant_pattern_matching",
         "this lint has been changed to redundant_pattern_matching",
```
Comment (flip1995) [inline clippy_lints/src/lib.rs]: That would've been better as `registered_renamed`, but that probably didn't exist in Clippy back then

**Comment by flip1995** [comment]
> I haven't tried it, but wouldn't it be enough to register both the tool version and the non-tool version as removed (on the Clippy side)?

We can try it. But that should only be required for lints that were deprecated before tool lints existed. So I would place them in a separate function or write a `deprecate_old!` macro to do this.

**Comment by phansch** [comment]
@flip1995 should be good to go now

File: tests/ui/deprecated.rs [hunk 3]
```diff
@@ -3,5 +3,6 @@
 #[warn(unstable_as_slice)]
 #[warn(unstable_as_mut_slice)]
 #[warn(misaligned_transmute)]
+#[warn(unused_collect)]
```
Comment (flip1995) [inline tests/ui/deprecated.rs]: I think we should move this test to `deprecated_old.rs` and create a new test file for tool lints deprecation.

**Comment by flip1995** [comment]
Thanks!

@bors r+

**Comment by matthiaskrgr** [comment]
Looks spurious?
````
   Compiling cargo_metadata v0.8.1

   Compiling clippy_lints v0.0.212 (C:\Users\travis\build\rust-lang\rust-clippy\clippy_lints)

    Finished dev [unoptimized + debuginfo] target(s) in 4m 27s

+++ cargo test --features 'debugging deny-warnings'

error: unable to unlink old fallback exe

info: caused by: Access is denied. (os error 5)

info: backtrace:

stack backtrace:

   0: <no info> (0x72aa68)

   1: <no info> (0x72aeca)

   2: <no info> (0x7275db)

   3: <no info> (0x5511eb)

   4: <no info> (0x5b9049)

   5: <no info> (0x470d5c)

   6: <no info> (0x46d302)

   7: <no info> (0x407bb6)

   8: <no info> (0x473032)

   9: <no info> (0x4013f8)

  10: <no info> (0x40151b)

  11: BaseThreadInitThunk (0x7ffcc8f03034)
````
@bors retry

**Pull request merged**